### PR TITLE
ARC-0012 Name Service

### PR DESCRIPTION
**Pull request by barnjamin**
Very rough first draft for a Name Service

### CODE REVIEW

**Comment by emg110** [comment]
Since Algorand has the naming infrastructure essentials (Private / Human Readable Public keys), IMHO,  it's a very important requirement to implement small remaining standardization, conventions, and implementations to have NS dynamic linking and the ability to use even more human-readable and shorter names. Making it DoS resistant would be paramount. I think this is a wonderful proposition.

**Comment by barnjamin** [comment]
Also considering allowing something like `silvio.algo/twitter` where `silvio.algo` is the name and `/twitter` refers to a specific property rather than `twitter.silvio.algo`

File: ARCs/arc-0012.md [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+
+- The key of a property MUST be a byte string
+- The length in bytes of a property key/value pair MUST be <= 128 bytes.
+
+
+# Contracts
+
+A Stateless Contract (Name Record) is used as a contract account. The logic is a standadized Template Contract with one Template Variable, the Name. This provides a predictable address to use for resolution.  Local State for the address is used to store arbitrary Properties that may be resolved for the name.  
+
+A Stateful Contract (Name Registry) validates the Name Record contract and allows storage of Properties assocaited with a Name to be stored in its local state.
```
Comment (jannotti) [inline ARCs/arc-0012.md:82]: "associated"

File: ARCs/arc-0012.md [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+- A Name Registry MUST allow the Name Record to deregister if the transaction is co-signed by the Owner
+
+
+### Name Register
+Initializing the Properties for a Name.
+
+To register a Name Record an Application OptIn transaction is submitted where:
+
+1. The ApplicationArgs array is populated with the Name Record bytecode and properties ([bytecode, key1, value1, key2, value2 ,...])
+2. The Account array is populated with the Name Record Account annd the registering Owner
```
Comment (jannotti) [inline ARCs/arc-0012.md:106]: annd -> and

**Comment by chicopaec** [comment]
@barnjamin 
> Also considering allowing something like silvio.algo/twitter where silvio.algo is the name and /twitter refers to a specific property rather than twitter.silvio.algo

What is the benefit over introducing a new path syntax that eschews the typical hierarchy of URLs? Using paths would be "$ID . $ROOT / $SUBID" whereas keeping with the traditional system would be "$SUBID . $ID . $ROOT". Not only more consistent, but also easier for people to learn.

**Pull request closed without merging**